### PR TITLE
feat: support custom dialer

### PR DESCRIPTION
**Pull request by themihai**

### CODE REVIEW

File: memcache/memcache.go [hunk 1]
```diff
@@ -258,8 +260,10 @@ func (c *Client) dial(addr net.Addr) (net.Conn, error) {
 		cn  net.Conn
 		err error
 	}
-
-	nc, err := net.DialTimeout(addr.Network(), addr.String(), c.netTimeout())
+	if c.DialTimeout == nil {
+		c.DialTimeout = net.DialTimeout
```
Comment (bboreham) [inline memcache/memcache.go:264]: I think assigning to the `Client` here is bad form - better to use a local variable, or add a function similar to `netTimeout()`

**Comment by bboreham** [comment]
This is broadly the same as #59

File: memcache/memcache.go [hunk 2]
```diff
@@ -258,8 +260,10 @@ func (c *Client) dial(addr net.Addr) (net.Conn, error) {
 		cn  net.Conn
 		err error
 	}
-
-	nc, err := net.DialTimeout(addr.Network(), addr.String(), c.netTimeout())
+	if c.DialTimeout == nil {
```
Comment (mar4uk) [inline memcache/memcache.go:263]: @themihai what do you think about @bboreham proposal below?
This proposal is already implemented in grafana fork of gomemcache: https://github.com/grafana/gomemcache/pull/2/files

```suggestion
        dialTimeout := c.DialTimeout
	if dialTimeout == nil {
		dialTimeout = net.DialTimeout
	}
	nc, err := dialTimeout(addr.Network(), addr.String(), c.netTimeout())
```
What do you think about adopting it here?
@bboreham I hope you don't mind I just copied your code for suggestion :)

File: memcache/memcache.go [hunk 3]
```diff
@@ -258,8 +260,10 @@ func (c *Client) dial(addr net.Addr) (net.Conn, error) {
...
+		c.DialTimeout = net.DialTimeout
+	}
+	nc, err := c.DialTimeout(addr.Network(), addr.String(), c.netTimeout())
```
Comment (mar4uk) [inline memcache/memcache.go:266]: @themihai what do you think about @bboreham proposal below?
This proposal is already implemented in grafana fork of gomemcache: https://github.com/grafana/gomemcache/pull/2/files

```suggestion
        dialTimeout := c.DialTimeout
	if dialTimeout == nil {
		dialTimeout = net.DialTimeout
	}
	nc, err := dialTimeout(addr.Network(), addr.String(), c.netTimeout())
```
What do you think about adopting it here?
@bboreham I hope you don't mind I just copied your code for the suggestion :)

**Comment by bradfitz** [comment]
Done in #158

**Pull request closed without merging**